File: arkanoid.rb
# arkanoid.rb
require 'io/console'
require 'timeout'

W = 40
H = 20

$paddle_x = W/2 - 4
$paddle_width = 8
$ball_x = $paddle_x + $paddle_width/2
$ball_y = H - 3
$ball_dx = 1
$ball_dy = -1
$ball_launched = false
$lives = 3
$score = 0
$level = 1
$bricks = []
$game_over = false
$running = true

def init_bricks
  $bricks = []
  rows = [3 + $level, 6].min
  cols = W / 3
  (0...rows).each do |y|
    (1...cols).each do |x|
      $bricks << {x: x*3, y: y+2, alive: true}
    end
  end
end

def clear_screen
  system('clear') || system('cls')
end

def draw_border
  (0...W).each { |x| print "\e[0;#{x}H#" }
  (0...W).each { |x| print "\e[#{H-1};#{x}H#" }
  (0...H).each { |y| print "\e[#{y};0H#" }
  (0...H).each { |y| print "\e[#{y};#{W-1}H#" }
end

def draw_paddle
  (0...$paddle_width).each { |x| print "\e[#{H-2};#{$paddle_x+x}H=" }
end

def draw_ball
  print "\e[#{$ball_y};#{$ball_x}HO" if $ball_launched || !$game_over
end

def draw_bricks
  $bricks.each do |b|
    print "\e[#{b[:y]};#{b[:x]}H█" if b[:alive]
  end
end

def draw_info
  print "\e[0;#{W+2}HScore: #{$score}"
  print "\e[1;#{W+2}HLives: #{$lives}"
  print "\e[2;#{W+2}HLevel: #{$level}"
  if $game_over
    print "\e[#{H/2};#{W/2-5}HGAME OVER! Press R"
  elsif !$ball_launched
    print "\e[#{H/2};#{W/2-5}HPress Space to launch"
  end
end

def update_ball
  return if !$ball_launched || $game_over
  $ball_x += $ball_dx
  $ball_y += $ball_dy
  # walls
  $ball_dx *= -1 if $ball_x <= 1 || $ball_x >= W-2
  $ball_dy *= -1 if $ball_y <= 1
  # paddle
  if $ball_y == H-2 && $ball_x >= $paddle_x && $ball_x < $paddle_x + $paddle_width
    $ball_dy *= -1
    hit_pos = $ball_x - ($paddle_x + $paddle_width/2)
    if hit_pos != 0
      $ball_dx = hit_pos > 0 ? 1 : -1
    end
  end
  # bricks
  $bricks.each do |b|
    if b[:alive] && $ball_y == b[:y] && $ball_x == b[:x]
      b[:alive] = false
      $score += 10
      $ball_dy *= -1
      break
    end
  end
  # bottom
  if $ball_y >= H-1
    $lives -= 1
    $ball_launched = false
    $ball_x = $paddle_x + $paddle_width/2
    $ball_y = H-3
    $ball_dx = rand(2) == 0 ? 1 : -1
    $ball_dy = -1
    $game_over = true if $lives == 0
  end
  # level complete
  if $bricks.all? { |b| !b[:alive] }
    $level += 1
    init_bricks
    $ball_launched = false
    $ball_x = $paddle_x + $paddle_width/2
    $ball_y = H-3
    $ball_dx = rand(2) == 0 ? 1 : -1
    $ball_dy = -1
  end
end

def move_paddle(dx)
  return if $game_over
  new_x = $paddle_x + dx
  if new_x >= 1 && new_x + $paddle_width <= W-1
    $paddle_x = new_x
    $ball_x = $paddle_x + $paddle_width/2 unless $ball_launched
  end
end

def launch_ball
  return if $game_over
  $ball_launched = true unless $ball_launched
end

def restart
  $paddle_x = W/2 - 4
  $ball_x = $paddle_x + $paddle_width/2
  $ball_y = H-3
  $ball_dx = 1
  $ball_dy = -1
  $ball_launched = false
  $lives = 3
  $score = 0
  $level = 1
  $game_over = false
  init_bricks
end

def input_loop
  while $running
    char = STDIN.getch
    case char
    when 'q', 'Q' then $running = false
    when 'r', 'R'
      restart if $game_over
    when ' ' then launch_ball
    when 'a', 'A' then move_paddle(-2)
    when 'd', 'D' then move_paddle(2)
    when "\e" # arrow keys
      c = STDIN.read_nonblock(2) rescue nil
      if c == '[D' then move_paddle(-2)
      elsif c == '[C' then move_paddle(2)
      end
    end
  end
end

init_bricks
Thread.new { input_loop }
while $running
  update_ball
  clear_screen
  draw_border
  draw_paddle
  draw_ball
  draw_bricks
  draw_info
  sleep 0.05
end

File: arkanoid.swift
// arkanoid.swift
import Foundation

let W = 40
let H = 20

var paddleX = W/2 - 4
let paddleWidth = 8
var ballX = paddleX + paddleWidth/2
var ballY = H - 3
var ballDX = 1
var ballDY = -1
var ballLaunched = false
var lives = 3
var score = 0
var level = 1
var bricks: [(x: Int, y: Int, alive: Bool)] = []
var gameOver = false
var running = true

func initBricks() {
    bricks = []
    let rows = min(3 + level, 6)
    let cols = W / 3
    for y in 0..<rows {
        for x in 1..<cols {
            bricks.append((x: x*3, y: y+2, alive: true))
        }
    }
}

func clearScreen() {
    print("\u{001B}[2J")
}

func drawBorder() {
    for x in 0..<W {
        print("\u{001B}[0;\(x)H#", terminator: "")
        print("\u{001B}[\(H-1);\(x)H#", terminator: "")
    }
    for y in 0..<H {
        print("\u{001B}[\(y);0H#", terminator: "")
        print("\u{001B}[\(y);\(W-1)H#", terminator: "")
    }
}

func drawPaddle() {
    for x in 0..<paddleWidth {
        print("\u{001B}[\(H-2);\(paddleX+x)H=", terminator: "")
    }
}

func drawBall() {
    if ballLaunched || !gameOver {
        print("\u{001B}[\(ballY);\(ballX)HO", terminator: "")
    }
}

func drawBricks() {
    for b in bricks {
        if b.alive {
            print("\u{001B}[\(b.y);\(b.x)H█", terminator: "")
        }
    }
}

func drawInfo() {
    print("\u{001B}[0;\(W+2)HScore: \(score)", terminator: "")
    print("\u{001B}[1;\(W+2)HLives: \(lives)", terminator: "")
    print("\u{001B}[2;\(W+2)HLevel: \(level)", terminator: "")
    if gameOver {
        print("\u{001B}[\(H/2);\(W/2-5)HGAME OVER! Press R", terminator: "")
    } else if !ballLaunched {
        print("\u{001B}[\(H/2);\(W/2-5)HPress Space to launch", terminator: "")
    }
}

func updateBall() {
    if !ballLaunched || gameOver { return }
    ballX += ballDX
    ballY += ballDY
    if ballX <= 1 || ballX >= W-2 { ballDX *= -1 }
    if ballY <= 1 { ballDY *= -1 }
    if ballY == H-2 && ballX >= paddleX && ballX < paddleX + paddleWidth {
        ballDY *= -1
        let hitPos = ballX - (paddleX + paddleWidth/2)
        if hitPos != 0 {
            ballDX = hitPos > 0 ? 1 : -1
        }
    }
    for i in 0..<bricks.count {
        if bricks[i].alive && ballY == bricks[i].y && ballX == bricks[i].x {
            bricks[i].alive = false
            score += 10
            ballDY *= -1
            break
        }
    }
    if ballY >= H-1 {
        lives -= 1
        ballLaunched = false
        ballX = paddleX + paddleWidth/2
        ballY = H-3
        ballDX = Int.random(in: 0...1) == 0 ? 1 : -1
        ballDY = -1
        if lives == 0 { gameOver = true }
    }
    if bricks.allSatisfy({ !$0.alive }) {
        level += 1
        initBricks()
        ballLaunched = false
        ballX = paddleX + paddleWidth/2
        ballY = H-3
        ballDX = Int.random(in: 0...1) == 0 ? 1 : -1
        ballDY = -1
    }
}

func movePaddle(dx: Int) {
    if gameOver { return }
    let newX = paddleX + dx
    if newX >= 1 && newX + paddleWidth <= W-1 {
        paddleX = newX
        if !ballLaunched {
            ballX = paddleX + paddleWidth/2
        }
    }
}

func launchBall() {
    if gameOver { return }
    if !ballLaunched { ballLaunched = true }
}

func restart() {
    paddleX = W/2 - 4
    ballX = paddleX + paddleWidth/2
    ballY = H-3
    ballDX = 1
    ballDY = -1
    ballLaunched = false
    lives = 3
    score = 0
    level = 1
    gameOver = false
    initBricks()
}

func inputLoop() {
    while running {
        let input = readLine(strippingNewline: false) ?? ""
        let chars = Array(input)
        if chars.isEmpty { continue }
        let ch = chars[0]
        switch ch {
        case "q", "Q": running = false; return
        case "r", "R":
            if gameOver { restart() }
        case " ":
            launchBall()
        case "a", "A": movePaddle(dx: -2)
        case "d", "D": movePaddle(dx: 2)
        default: break
        }
    }
}

initBricks()
DispatchQueue.global().async {
    inputLoop()
}
while running {
    updateBall()
    clearScreen()
    drawBorder()
    drawPaddle()
    drawBall()
    drawBricks()
    drawInfo()
    Thread.sleep(forTimeInterval: 0.05)
}
